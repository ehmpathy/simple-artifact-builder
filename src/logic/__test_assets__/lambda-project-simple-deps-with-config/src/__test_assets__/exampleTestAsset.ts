export const asset = () => console.log('hello');
