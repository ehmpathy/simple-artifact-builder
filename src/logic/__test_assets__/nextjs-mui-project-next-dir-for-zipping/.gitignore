.artifact/contents.zip
.artifact/contents.sizes.json
