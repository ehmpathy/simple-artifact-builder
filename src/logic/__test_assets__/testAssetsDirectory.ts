export const TEST_ASSETS_DIRECTORY = __dirname;
