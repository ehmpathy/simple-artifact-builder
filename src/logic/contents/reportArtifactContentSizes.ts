import fs from 'fs';
import prettyBytes from 'pretty-bytes';

import { assertThatArtifactContentsWereBuilt } from './assertThatArtifactContentsWereBuilt';

const roundToHundredths = (x: number) => Math.round(100 * x) / 100;

export enum DirectoryContentType {
  FILE = 'FILE',
  DIRECTORY = 'DIRECTORY',
}
interface DirectoryContent {
  path: string;
  type: DirectoryContentType;
  sizeInBytes: number | null; // null for directory
}

const getDirectoryContents = async ({
  absoluteDirectoryPath,
}: {
  absoluteDirectoryPath: string;
}): Promise<DirectoryContent[]> => {
  const names = await fs.promises.readdir(absoluteDirectoryPath);
  const stats = await Promise.all(
    names.map(async (name) => ({
      name,
      path: `${absoluteDirectoryPath}/${name}`,
      stat: await fs.promises.stat(`${absoluteDirectoryPath}/${name}`),
    })),
  );
  return stats.map(
    ({ path, stat }): DirectoryContent => ({
      path,
      type: stat.isDirectory()
        ? DirectoryContentType.DIRECTORY
        : DirectoryContentType.FILE,
      sizeInBytes: stat.isDirectory() ? null : stat.size,
    }),
  );
};
const getDirectoryContentsRecursively = async ({
  absoluteDirectoryPath,
}: {
  absoluteDirectoryPath: string;
}): Promise<DirectoryContent[]> => {
  // lookup the contents of this directory
  const contents = await getDirectoryContents({ absoluteDirectoryPath });

  // pick out the file contents, since this already have a size and dont need further processing
  const immediatelyNestedFiles = contents.filter(
    (content) => content.type === DirectoryContentType.FILE,
  );

  // now get the contents of each nested directory + define a size for each of these directories based on their immediate children
  const immediatelyNestedDirectoriesWithoutSizes = contents.filter(
    (content) => content.type === DirectoryContentType.DIRECTORY,
  );
  const immediatelyNestedDirectoriesAndTheirRecursivelyNestedContentsWithSizes =
    (
      await Promise.all(
        immediatelyNestedDirectoriesWithoutSizes.map((nestedDirectory) =>
          getDirectoryContentsRecursively({
            absoluteDirectoryPath: nestedDirectory.path,
          }),
        ),
      )
    ).flat();

  // now define the immediate children of this directory
  const immediateContentsWithSizes = [
    ...immediatelyNestedFiles,
    ...immediatelyNestedDirectoriesAndTheirRecursivelyNestedContentsWithSizes.filter(
      (content) =>
        content.path.split('/').slice(0, -1).join('/') ===
        absoluteDirectoryPath, // the directory that content is in === this directory
    ),
  ];

  // now define the size of this directory
  const thisDirectoryContent: DirectoryContent = {
    path: absoluteDirectoryPath,
    type: DirectoryContentType.DIRECTORY,
    sizeInBytes: immediateContentsWithSizes.reduce(
      (summary, thisContentFile) => summary + thisContentFile.sizeInBytes!,
      0,
    ), // add up the sizes of the immediate children
  };

  // now merge the contents and return them
  return [
    thisDirectoryContent,
    ...immediatelyNestedFiles,
    ...immediatelyNestedDirectoriesAndTheirRecursivelyNestedContentsWithSizes,
  ];
};

export const reportArtifactContentSizes = async ({
  projectRootDirectory,
}: {
  projectRootDirectory: string;
}) => {
  const toAbsolutePath = (relativePath: string) =>
    `${projectRootDirectory}/${relativePath}`;

  // check that contents to get stats about exist
  await assertThatArtifactContentsWereBuilt({ projectRootDirectory });

  // get file size of total contents directory
  const directoryContentsWithAbsolutePaths =
    await getDirectoryContentsRecursively({
      absoluteDirectoryPath: toAbsolutePath('.artifact/contents'),
    });

  // drop the project root directory from the paths
  const directoryContents = directoryContentsWithAbsolutePaths.map(
    (content) => ({
      ...content,
      path: content.path
        .replace(`${projectRootDirectory}/.artifact/contents`, '')
        .replace(/^\//, '')
        .replace(/^$/, '.'), // at this point the root directories path is an empty string. update that to `.`
    }),
  );

  // sort the directory contents by size
  const sortedDirectoryContents = directoryContents.sort((a, b) =>
    a.sizeInBytes! > b.sizeInBytes! ? -1 : 1,
  );

  // and make the sizes human readable
  const totalSize = sortedDirectoryContents[0]?.sizeInBytes!;
  const sortedDirectoryContentsWithHumanReadableSizes =
    sortedDirectoryContents.map((content) => ({
      ...content,
      sizeInHuman: prettyBytes(content.sizeInBytes!),
      sizeInScale: `${roundToHundredths(
        (content.sizeInBytes! / totalSize) * 100,
      )}%`,
    }));

  // record the contents "manifest" (i.e., all of the files that are in the contents)
  await fs.promises.writeFile(
    toAbsolutePath('.artifact/contents.sizes.json'),
    JSON.stringify(
      { sizes: sortedDirectoryContentsWithHumanReadableSizes },
      null,
      2,
    ),
  );
};
