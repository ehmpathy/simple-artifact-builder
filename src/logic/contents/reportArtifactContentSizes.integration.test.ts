import fs from 'fs';
import { TEST_ASSETS_DIRECTORY } from '../__test_assets__/testAssetsDirectory';
import { reportArtifactContentSizes } from './reportArtifactContentSizes';

describe('reportArtifactContentsSize', () => {
  it('should be able to report the size of an artifact contents directory, with total + all files/dirs over 5% size showing', async () => {
    const projectRootDirectory = `${TEST_ASSETS_DIRECTORY}/nextjs-mui-project-next-dir-for-sizing`;
    const targetOutputFilePath = `${projectRootDirectory}/.artifact/contents.sizes.json`;

    // remove the previous file if it exists
    await fs.promises.unlink(targetOutputFilePath).catch(() => {}); // do nothing on error, e.g., if file didn't already exist

    // generate the file
    await reportArtifactContentSizes({ projectRootDirectory });

    // check its contents
    const contents = JSON.parse((await fs.promises.readFile(targetOutputFilePath, { encoding: 'utf8' })) as string) as {
      sizes: any[];
    };
    expect(contents.sizes.length).toBeGreaterThan(1);
    expect(contents).toMatchSnapshot(); // log the example
  });
});
