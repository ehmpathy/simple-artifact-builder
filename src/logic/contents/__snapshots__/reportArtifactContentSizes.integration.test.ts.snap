// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reportArtifactContentsSize should be able to report the size of an artifact contents directory, with total + all files/dirs over 5% size showing 1`] = `
Object {
  "sizes": Array [
    Object {
      "path": ".",
      "sizeInBytes": 973745,
      "sizeInHuman": "974 kB",
      "sizeInScale": "100%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next",
      "sizeInBytes": 973298,
      "sizeInHuman": "973 kB",
      "sizeInScale": "99.95%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/static",
      "sizeInBytes": 416158,
      "sizeInHuman": "416 kB",
      "sizeInScale": "42.74%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/static/chunks",
      "sizeInBytes": 415536,
      "sizeInHuman": "416 kB",
      "sizeInScale": "42.67%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/server",
      "sizeInBytes": 314598,
      "sizeInHuman": "315 kB",
      "sizeInScale": "32.31%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/trace",
      "sizeInBytes": 224537,
      "sizeInHuman": "225 kB",
      "sizeInScale": "23.06%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages",
      "sizeInBytes": 167479,
      "sizeInHuman": "167 kB",
      "sizeInScale": "17.2%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/server/chunks",
      "sizeInBytes": 132543,
      "sizeInHuman": "133 kB",
      "sizeInScale": "13.61%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/static/chunks/framework-bbce3cdc1a611f18.js",
      "sizeInBytes": 131009,
      "sizeInHuman": "131 kB",
      "sizeInScale": "13.45%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/main-e7f9cd3517807e31.js",
      "sizeInBytes": 95171,
      "sizeInHuman": "95.2 kB",
      "sizeInScale": "9.77%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/polyfills-a40ef1678bae11e696dba45124eadd70.js",
      "sizeInBytes": 91434,
      "sizeInHuman": "91.4 kB",
      "sizeInScale": "9.39%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/chunks/664.js",
      "sizeInBytes": 70910,
      "sizeInHuman": "70.9 kB",
      "sizeInScale": "7.28%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/pages",
      "sizeInBytes": 48027,
      "sizeInHuman": "48 kB",
      "sizeInScale": "4.93%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/server/chunks/859.js",
      "sizeInBytes": 40411,
      "sizeInHuman": "40.4 kB",
      "sizeInScale": "4.15%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/pages/_app-02318ea03113095b.js",
      "sizeInBytes": 38569,
      "sizeInHuman": "38.6 kB",
      "sizeInScale": "3.96%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/755-1f25bec492800030.js",
      "sizeInBytes": 30519,
      "sizeInHuman": "30.5 kB",
      "sizeInScale": "3.13%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/_document.js.nft.json",
      "sizeInBytes": 28399,
      "sizeInHuman": "28.4 kB",
      "sizeInScale": "2.92%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/about.js.nft.json",
      "sizeInBytes": 27473,
      "sizeInHuman": "27.5 kB",
      "sizeInScale": "2.82%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/index.js.nft.json",
      "sizeInBytes": 22987,
      "sizeInHuman": "23 kB",
      "sizeInScale": "2.36%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/_app.js.nft.json",
      "sizeInBytes": 20747,
      "sizeInHuman": "20.7 kB",
      "sizeInScale": "2.13%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/31-ea2e06847afa5ee5.js",
      "sizeInBytes": 17904,
      "sizeInHuman": "17.9 kB",
      "sizeInScale": "1.84%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/about.html",
      "sizeInBytes": 15733,
      "sizeInHuman": "15.7 kB",
      "sizeInScale": "1.62%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/index.html",
      "sizeInBytes": 13447,
      "sizeInHuman": "13.4 kB",
      "sizeInScale": "1.38%",
      "type": "FILE",
    },
    Object {
      "path": ".next/next-server.js.nft.json",
      "sizeInBytes": 13395,
      "sizeInHuman": "13.4 kB",
      "sizeInScale": "1.38%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/chunks/600.js",
      "sizeInBytes": 11653,
      "sizeInHuman": "11.7 kB",
      "sizeInScale": "1.2%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/404.html",
      "sizeInBytes": 10992,
      "sizeInHuman": "11 kB",
      "sizeInScale": "1.13%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/500.html",
      "sizeInBytes": 10978,
      "sizeInHuman": "11 kB",
      "sizeInScale": "1.13%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/font-manifest.json",
      "sizeInBytes": 8243,
      "sizeInHuman": "8.24 kB",
      "sizeInScale": "0.85%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/chunks/font-manifest.json",
      "sizeInBytes": 8243,
      "sizeInHuman": "8.24 kB",
      "sizeInScale": "0.85%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/_document.js",
      "sizeInBytes": 7151,
      "sizeInHuman": "7.15 kB",
      "sizeInScale": "0.73%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/webpack-runtime.js",
      "sizeInBytes": 6062,
      "sizeInHuman": "6.06 kB",
      "sizeInScale": "0.62%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/_app.js",
      "sizeInBytes": 5097,
      "sizeInHuman": "5.1 kB",
      "sizeInScale": "0.52%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/pages/about-a0e709c255002672.js",
      "sizeInBytes": 4617,
      "sizeInHuman": "4.62 kB",
      "sizeInScale": "0.47%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/pages/index-259f2ab4bd6663c8.js",
      "sizeInBytes": 4574,
      "sizeInHuman": "4.57 kB",
      "sizeInScale": "0.47%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/_error.js",
      "sizeInBytes": 3830,
      "sizeInHuman": "3.83 kB",
      "sizeInScale": "0.39%",
      "type": "FILE",
    },
    Object {
      "path": ".next/required-server-files.json",
      "sizeInBytes": 2020,
      "sizeInHuman": "2.02 kB",
      "sizeInScale": "0.21%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/webpack-514908bffb652963.js",
      "sizeInBytes": 1472,
      "sizeInHuman": "1.47 kB",
      "sizeInScale": "0.15%",
      "type": "FILE",
    },
    Object {
      "path": ".next/build-manifest.json",
      "sizeInBytes": 1417,
      "sizeInHuman": "1.42 kB",
      "sizeInScale": "0.15%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/chunks/730.js",
      "sizeInBytes": 1326,
      "sizeInHuman": "1.33 kB",
      "sizeInScale": "0.14%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages/_error.js.nft.json",
      "sizeInBytes": 645,
      "sizeInHuman": "645 B",
      "sizeInScale": "0.07%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/lwiVzzFlheitLtrmf0quQ",
      "sizeInBytes": 622,
      "sizeInHuman": "622 B",
      "sizeInScale": "0.06%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/static/lwiVzzFlheitLtrmf0quQ/_buildManifest.js",
      "sizeInBytes": 453,
      "sizeInHuman": "453 B",
      "sizeInScale": "0.05%",
      "type": "FILE",
    },
    Object {
      "path": "dist",
      "sizeInBytes": 447,
      "sizeInHuman": "447 B",
      "sizeInScale": "0.05%",
      "type": "DIRECTORY",
    },
    Object {
      "path": ".next/routes-manifest.json",
      "sizeInBytes": 423,
      "sizeInHuman": "423 B",
      "sizeInScale": "0.04%",
      "type": "FILE",
    },
    Object {
      "path": ".next/images-manifest.json",
      "sizeInBytes": 321,
      "sizeInHuman": "321 B",
      "sizeInScale": "0.03%",
      "type": "FILE",
    },
    Object {
      "path": ".next/prerender-manifest.json",
      "sizeInBytes": 312,
      "sizeInHuman": "312 B",
      "sizeInScale": "0.03%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/chunks/pages/_error-2280fa386d040b66.js",
      "sizeInBytes": 267,
      "sizeInHuman": "267 B",
      "sizeInScale": "0.03%",
      "type": "FILE",
    },
    Object {
      "path": "dist/handler.js",
      "sizeInBytes": 259,
      "sizeInHuman": "259 B",
      "sizeInScale": "0.03%",
      "type": "FILE",
    },
    Object {
      "path": "dist/server.js",
      "sizeInBytes": 188,
      "sizeInHuman": "188 B",
      "sizeInScale": "0.02%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/pages-manifest.json",
      "sizeInBytes": 187,
      "sizeInHuman": "187 B",
      "sizeInScale": "0.02%",
      "type": "FILE",
    },
    Object {
      "path": ".next/export-marker.json",
      "sizeInBytes": 94,
      "sizeInHuman": "94 B",
      "sizeInScale": "0.01%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/lwiVzzFlheitLtrmf0quQ/_middlewareManifest.js",
      "sizeInBytes": 92,
      "sizeInHuman": "92 B",
      "sizeInScale": "0.01%",
      "type": "FILE",
    },
    Object {
      "path": ".next/server/middleware-manifest.json",
      "sizeInBytes": 84,
      "sizeInHuman": "84 B",
      "sizeInScale": "0.01%",
      "type": "FILE",
    },
    Object {
      "path": ".next/static/lwiVzzFlheitLtrmf0quQ/_ssgManifest.js",
      "sizeInBytes": 77,
      "sizeInHuman": "77 B",
      "sizeInScale": "0.01%",
      "type": "FILE",
    },
    Object {
      "path": ".next/BUILD_ID",
      "sizeInBytes": 21,
      "sizeInHuman": "21 B",
      "sizeInScale": "0%",
      "type": "FILE",
    },
    Object {
      "path": ".next/react-loadable-manifest.json",
      "sizeInBytes": 2,
      "sizeInHuman": "2 B",
      "sizeInScale": "0%",
      "type": "FILE",
    },
  ],
}
`;
